src/compositor.rs

```diff
@@ -401,3 +407,27 @@ pub fn terminate() {
         }
     }
 }
+
+/// Runs a function with the compositor as an argument.
+///
+/// Note this will fail if you are within a wayland callback because at that
+/// time the compositor is provided to you.
+///
+/// This method is only useful if you are trying to operate on the compositor
+/// state while not in the wayland event loop.
+pub fn with_compositor<F, R>(func: F) -> HandleResult<R>
+    where F: FnOnce(&mut Compositor) -> R
+{
+    unsafe {
+        if COMPOSITOR_PTR.is_null() {
+            return Err(HandleErr::AlreadyDropped)
+        }
+        let compositor = &mut *COMPOSITOR_PTR;
+        if compositor.lock.get() {
+            return Err(HandleErr::AlreadyBorrowed)
+        } else {
+            compositor.lock.set(true);
+            Ok(func(compositor))
```

Comment:
I might be missing something, but where is the lock set to `false` again?
(Note to self: It would be nice to provide a wrapper for the above; i.e. some kind of `.get()` function which returns either an `Err(HandleErr::AlreadyBorrowed)` or an `Ok(LockHandle)`, where `LockHandle` implements `DerefMut<TargetType>` and in its `Drop` implementation it releases the lock; i.e. similar to how a `MutexGuard` operates)

Reply:
Woops! Nice catch.

> It would be nice to provide a wrapper for the above; 

Unfortunately... there are certain use cases that require recursively locking a handle and if you set it back to `false` in that inner lock then you can suddenly lock it again.


An example: when you add an `Output` to an `OutputLayout` then `OutputLayoutHandler::output_added` triggers. This borrows the `OutputLayout` from the handle. But it's already borrowed, since how else was it added in the first place? If we have a wrapper type that, on drop, sets it to be `false` then after that code we could borrow it again.

e.g here is it in code:


```rust
// Doesn't matter where these came from
let output_layout_handle: OutputLayoutHandle = ...;
let output_handle: OutputHandle = ...
run_handles!([(output: {output_handle.clone()}), (output_layout: {output_layout_handle.clone()})] => {
    // This triggers OutputLayoutHandler::output_added
    output_layout.add_auto(output);
    // If, naively, that set the handle back to false we can borrow the output_layout again...
    run_handles!([(output_layout: {output_layout})] => { /* boom */})
})
```

I fixed that bug in this patch actually.

Reply:
Or actually there's a different idea I have...


This problem could easily come up again but for a different type in wlroots-rs. It's very easy for the callstack to get very hairy when calling from a wlroots-rs callabck and for there to be an attempt to borrow a handle when it's already borrowed with no way to pass it down. To make that worse, you can't simply `drop` the type since it's borrowed and the `false` flag won't actually be set until afterwards manually!

*and* those manual flag setting is error-prone...as seen here.


So why not instead _all_ callbacks take `Handle` types instead that you have to upgrade explicitly? It adds some slight boiler plate, but thanks to `run_handles!` (which should be renamed to `with_handles!`) it's not that bad. And it gives *far more* control to the compositor writer.

That will mean I need to return a `CompositorHandle` then instead of a `Compositor`, which is even more breakage but something I can live with since this is going to break all of the examples and Way Cooler code.